### ISSUE TITLE
Bug - model trained on X from two sample texts

### ISSUE THREAD

**Issue by Majdoddin** [karpathy/nanoGPT#320]
Each sample of openwebtext consists of several paragraphs extracted from a single webpage. nanoGPT is trained to predict a token of a sample, given its previous tokens. For train split these texts are concatenated (in prepare.py) and mapped to memory as a large array of tokens.
https://github.com/karpathy/nanoGPT/blob/eba36e84649f3c6d840a93092cb779a260544d08/train.py#L116
During the batch sampling process, each sample is a segment of this token array with a length specified as block_size, starting from a random point. However, a problem arises when the segment extends beyond the end of one text and into the beginning of the next text. 
https://github.com/karpathy/nanoGPT/blob/eba36e84649f3c6d840a93092cb779a260544d08/train.py#L120
This issue distorts the training because consecutive texts are typically unrelated, and predicting the next text based on the previous text does not make sense.

This bug is subtle to notice, because the same issue exists in the evaluation batches.
If confirmed, I can write a PR to fix the issue.

**Comment by jyegerlehner** [comment]
I think this is a valid point. However, there is an end-of-text (`eot_token`) that marks the transition between sources, which is added in `prepare.py`. So I think that means we are telling the LLM "we are changing topics now; the text before the `eot_token` is unrelated to the text after the `eot_token`". Or rather giving it enough information to figure out that text before and after `eot_token` are unrelated.

This brings up a related issue: we are oftentimes telling the LLM that the correct next-token to predict is an end-of-text character. Is this what we want? There might not be enough context for the LLM to be able to predict where the text generation should end. And also, the LLM's prediction of `eot_token` is not being used to end text generation. Rather, a fixed number of characters is always being generated (I think, based on a glance at sample.py). 

So I am tempted to say `eot_token` as the next character in training data is pollution.  An approach to eliminate them is to take the random index produced train.py's `get_batch,` then search forward and backward for the 'eot_token' before and after that index, and use those as bounds between which we randomly select the block of text from. There is the added complication that sometimes the text between the bounding `eot_tokens` is less than `block_size,` but the tensor must be exactly block_size. In that case, what do we pad it with? -1 to tell the softmax to ignore it maybe?

**Comment by jyegerlehner** [comment]
So I did an experiment. I made changes to the data loader described above. So what's different about the behavior is the following:
1) To @Majdoddin 's point, a training example never presents text that is split across two sources. All the text in the block comes from between two `eot_tokens`.
2) The dataloader never presents `eot_token` as the next word. This is also a consequence of selecting the text from between two `eot_tokens.`
3) Single-source text in training set is often shorter than the gpt2 1024 context window, and in these cases the context is padded with the token for a space character (220, I think it turned out to be).

As a test, I trained a gpt2 model de novo using 
`python3 train.py ./config/train_gpt2.py`

using [this branch](https://github.com/jyegerlehner/nanoGPT/tree/eot_token-handling) and again using the master branch of nanoGPT.

![image](https://github.com/karpathy/nanoGPT/assets/2138320/211a3dbf-3c0a-46af-b813-9aaa7ade86c6)

So it is learning faster and better (per iter, not wall-clock faster), I think because of some combination of the effects we described above: we avoid mixing text from different sources, and we don't teach it to predict eot_token as next character.

The problems with the code are: it is ugly and verbose (could be some bugs) but worst of all it is slow. `train.py` spends about half it's time loading data, mostly I think because of the linear searching for `eot_tokens` and concatenating numpy arrays to do the padding. Training is about twice as slow as master branch on the 4090 GPU, Ryzen 5800X CPU. 

I think the way forward with this might be to do this processing in `prepare.py` so the preprocessing penalty only has to be paid once instead of at training time, unless someone finds a much more efficient way to get this done. One problem with that approach is that it requires knowing the `context_size` of the LLM at the time `prepare.py` is run, so  `prepare.py` would have to be re-run every time you change the `context_size.` 

@Majdoddin I'd like to see what your approach was to this if you want to share it.

**Comment by shendiaomo** [comment]
> So I did an experiment. I made [changes ](https://github.com/jyegerlehner/nanoGPT/tree/eot_token-handling) to the data loader described above. So what's different about the behavior is the following:
> 
> 1. To @Majdoddin 's point, a training example never presents text that is split across two sources. All the text in the block comes from between two `eot_tokens`.
> 2. The dataloader never presents `eot_token` as the next word. This is also a consequence of selecting the text from between two `eot_tokens.`
> 3. Single-source text in training set is often shorter than the gpt2 1024 context window (`block_size` in the code), and in these cases the context is padded with the token for a space character (220, I think it turned out to be).
> 
> As a test, I trained a gpt2 model de novo using `python3 train.py ./config/train_gpt2.py`
> 
> using [this branch](https://github.com/jyegerlehner/nanoGPT/tree/eot_token-handling) and again using the master branch of nanoGPT.
> 
> ![image](https://user-images.githubusercontent.com/2138320/260487255-211a3dbf-3c0a-46af-b813-9aaa7ade86c6.png)
> 
> So it is learning faster and better (per iter, not wall-clock faster), I think because of some combination of the effects we described above: we avoid mixing text from different sources, and we don't teach it to predict eot_token as next character.
> 
> The problems with the code are: it is ugly and verbose (could be some bugs) but worst of all it is slow. `train.py` spends about half it's time loading data, mostly I think because of the linear searching for `eot_tokens` and concatenating numpy arrays to do the padding. Training is about twice as slow as master branch on the 4090 GPU, Ryzen 5800X CPU.
> 
> I think the way forward with this might be to do this processing in `prepare.py` so the preprocessing penalty only has to be paid once instead of at training time, unless someone finds a much more efficient way to get this done. One problem with that approach is that it requires knowing the `block_size` of the LLM at the time `prepare.py` is run, so `prepare.py` would have to be re-run every time you change the `block_size.`
> 
> @Majdoddin I'd like to see what your approach was to this if you want to share it.

Any follow-up? I believe that `eot` could be well learned as @karpathy described in #244 .

**Comment by jyegerlehner** [comment]
@shendiaomo  I was slow to pick up on the fact this lib isn't being maintained, and there are other projects that are probably better to build with, at least beyond learning exercises.  [lit_gpt](https://github.com/Lightning-AI/lit-gpt), for example, appears to have been based on nanoGPT and is actively developed and problems of this kind have been worked out, with support built in for popular open datasets and models.

So in short: no i haven't done any more with this.

> I believe that eot could be well learned as @karpathy described in https://github.com/karpathy/nanoGPT/issues/244 .

Sure, probably so. Better if it doesn't have to, though, i think.

**Comment by Majdoddin** [comment]
@jyegerlehner This is indeed very interesting :+1:  Your training loss is significantly better. 
Sorry that I had missed your response.

I also don't think `eot` is a good solution. Firstly, the model should discover the semantic of `eot`, that is the text before and after `eot` are irrelevant. This is by no means easy (fake correlations, ...). If it could be left on the model to learn, then why couldn't the model learn causality instead of using masking?

Second, QKV matrices should be trained to ignore the tokens before `eot`, when processing the tokens after `eot` (positional encoding is added to the tokens). I am not sure if it is possible at all.
